cpu/sam0_common/periph/rtt.c

```diff
@@ -33,39 +33,96 @@ static void* _overflow_arg;
 static rtt_cb_t _cmp0_cb;
 static void* _cmp0_arg;
 
-void rtt_init(void)
+static void _wait_syncbusy(void)
 {
-    DEBUG("%s:%d\n", __func__, __LINE__);
-    /* Turn on power manager for RTC */
-    MCLK->APBAMASK.reg |= MCLK_APBAMASK_RTC | MCLK_APBAMASK_OSC32KCTRL;
-    rtt_poweron();
+#ifdef REG_RTC_MODE2_SYNCBUSY
+    while (RTC->MODE2.SYNCBUSY.reg) {}
+#else
+    while(RTC->MODE2.STATUS.bit.SYNCBUSY) {}
+#endif
+}
 
-    /* reset */
-    RTC->MODE0.CTRLA.bit.SWRST = 1;
-    while(RTC->MODE0.CTRLA.bit.SWRST) {}
+static inline void _rtc_set_enabled(bool on)
+{
+#ifdef REG_RTC_MODE0_CTRLA
+    RTC->MODE0.CTRLA.bit.ENABLE = on;
+#else
+    RTC->MODE0.CTRL.bit.ENABLE = on;
+#endif
+}
 
-    /* set 32bit counting mode */
-    RTC->MODE0.CTRLA.bit.MODE = 0;
+#ifdef CPU_SAMD21
+static void _rtt_clock_setup(void)
+{
+    /* RTC uses External 32,768KHz Oscillator because OSC32K isn't accurate
+     * enough (p1075/1138). Also keep running in standby. */
+    SYSCTRL->XOSC32K.reg =  SYSCTRL_XOSC32K_ONDEMAND |
+                            SYSCTRL_XOSC32K_EN32K |
+                            SYSCTRL_XOSC32K_XTALEN |
+                            SYSCTRL_XOSC32K_STARTUP(6) |
+#if RTT_RUNSTDBY
+                            SYSCTRL_XOSC32K_RUNSTDBY |
+#endif
+                            SYSCTRL_XOSC32K_ENABLE;
+
+    /* Setup clock GCLK2 with divider 1 */
+    GCLK->GENDIV.reg = GCLK_GENDIV_ID(2) | GCLK_GENDIV_DIV(1);
+    while (GCLK->STATUS.bit.SYNCBUSY) {}
+
+    /* Enable GCLK2 with XOSC32K as source. Use divider without modification
+     * and keep running in standby. */
+    GCLK->GENCTRL.reg = GCLK_GENCTRL_ID(2) |
+                        GCLK_GENCTRL_GENEN |
+#if RTT_RUNSTDBY
+                        GCLK_GENCTRL_RUNSTDBY |
+#endif
+                        GCLK_GENCTRL_SRC_XOSC32K;
+    while (GCLK->STATUS.bit.SYNCBUSY) {}
+
+    /* Connect GCLK2 to RTC */
+    GCLK->CLKCTRL.reg = GCLK_CLKCTRL_GEN_GCLK2 |
+                        GCLK_CLKCTRL_CLKEN |
+                        GCLK_CLKCTRL_ID(RTC_GCLK_ID);
+    while (GCLK->STATUS.bit.SYNCBUSY) {}
+}
+/* !CPU_SAMD21 */
+#else
+static void _rtt_clock_setup(void)
+{
+    /* Turn on power manager for RTC */
+    MCLK->APBAMASK.reg |= MCLK_APBAMASK_OSC32KCTRL;
 
     /* set clock source */
     OSC32KCTRL->RTCCTRL.reg = OSC32KCTRL_RTCCTRL_RTCSEL_ULP32K;
+}
+#endif
 
-    /* enable */
-    RTC->MODE0.CTRLA.bit.ENABLE = 1;
-    while(RTC->MODE0.SYNCBUSY.bit.ENABLE) {}
+void rtt_init(void)
+{
+    _rtt_clock_setup();
+    rtt_poweron();
+
+    /* set 32bit counting mode */
+#ifdef REG_RTC_MODE0_CTRLA
+    RTC->MODE0.CTRLA.reg = 0;
```

Comment:
Please use `RTC->MODE0.CTRLA.bit.MODE = 0;` instead or add a syncbusy right after.
A 32-bit write will trigger a ENABLE operation to disable this peripheral, sometimes the next call (`_rtc_set_enabled(1);`) is discarded because the operation isn't finished yet.
This causes some issue on SAML11. Sometimes the peripheral isn't enable so periph_rtt test don't work.

Reply:
Actually since both just write the CTRLA register, could we not just do `RTC->MODE0.CTRLA.reg = RTC_MODE0_CTRLA_ENABLE;` in one go?

Reply:
`MODE0.CTRLA.reg = RTC_MODE0_CTRLA_MODE(0) | RTC_MODE0_CTRLA_ENABLE;`
in one go so it is perfectly clear :)

Reply:
Alright!